### PR TITLE
[Serializer] Fix denormalization of arrays

### PR DESCRIPTION
**Pull request by dunglas**
| Q | A |
| --- | --- |
| Branch? | 3.1 |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #19545 |
| License | MIT |
| Doc PR | n/a |

ping @hensoko @theofidry

### CODE REVIEW

File: src/Symfony/Component/Serializer/Normalizer/ArrayDenormalizer.php [hunk 1]
```diff
@@ -46,12 +49,16 @@ public function denormalize($data, $class, $format = null, array $context = arra
...
-        return array_map(
-            function ($data) use ($serializer, $class, $format, $context) {
-                return $serializer->denormalize($data, $class, $format, $context);
-            },
-            $data
-        );
+        $builtinType = isset($context['key_type']) ? $context['key_type']->getBuiltinType() : null;
+        foreach ($data as $key => $value) {
+            if (null !== $builtinType && !call_user_func('is_'.$builtinType, $key)) {
+                throw new UnexpectedValueException(sprintf('The type of the key "%s" must be "%s" ("%s" given).', $key, $builtinType, gettype($key)));
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Normalizer/ArrayDenormalizer.php]: No need for the brackets, just a comma

File: src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php [hunk 2]
```diff
@@ -248,8 +248,22 @@ private function validateAndDenormalize($currentClass, $attribute, $data, $forma
...
-            $class = $type->getClassName();
+            if (
+                $type->isCollection() &&
+                null !== ($collectionValueType = $type->getCollectionValueType()) &&
+                Type::BUILTIN_TYPE_OBJECT === ($collectionValueBuiltinType = $collectionValueType->getBuiltinType())
+            ) {
+                $builtinType = Type::BUILTIN_TYPE_OBJECT;
+                $class = $collectionValueType->getClassName().'[]';
+
+                if (null !== ($collectionKeyType = $type->getCollectionKeyType())) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php]: Brackets not needed

File: src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php [hunk 3]
```diff
@@ -248,8 +248,22 @@ private function validateAndDenormalize($currentClass, $attribute, $data, $forma
                 return;
             }
 
-            $builtinType = $type->getBuiltinType();
-            $class = $type->getClassName();
+            if (
+                $type->isCollection() &&
+                null !== ($collectionValueType = $type->getCollectionValueType()) &&
+                Type::BUILTIN_TYPE_OBJECT === ($collectionValueBuiltinType = $collectionValueType->getBuiltinType())
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php]: Is the var $collectionValueBuiltinType used at all?

File: src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php [hunk 4]
```diff
@@ -248,8 +248,22 @@ private function validateAndDenormalize($currentClass, $attribute, $data, $forma
                 return;
             }
 
-            $builtinType = $type->getBuiltinType();
-            $class = $type->getClassName();
+            if (
+                $type->isCollection() &&
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php]: Condition on a single line

File: src/Symfony/Component/Serializer/Tests/Normalizer/ObjectNormalizerTest.php [hunk 5]
```diff
@@ -546,6 +557,21 @@ public function testRejectInvalidType()
...
+     * @expectedException UnexpectedValueException
+     * @expectedExceptionMessage The type of the key "a" must be "int" ("string" given).
+     */
+    public function testRejectInvalidKey()
+    {
+        $extractor = new PropertyInfoExtractor(array(), array(new PhpDocExtractor(), new ReflectionExtractor()));
+        $normalizer = new ObjectNormalizer(null, null, null, $extractor);
+        $serializer = new Serializer(array(new ArrayDenormalizer(), new DateTimeNormalizer(), $normalizer));
+
+        var_dump($serializer->denormalize(array(
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Tests/Normalizer/ObjectNormalizerTest.php]: Var_dump?

File: src/Symfony/Component/Serializer/composer.json [hunk 6]
```diff
@@ -24,9 +24,10 @@
         "symfony/property-access": "~2.8|~3.0",
         "symfony/http-foundation": "~2.8|~3.0",
         "symfony/cache": "~3.1",
-        "symfony/property-info": "~2.8|~3.0",
+        "symfony/property-info": "~3.1",
         "doctrine/annotations": "~1.0",
-        "doctrine/cache": "~1.0"
+        "doctrine/cache": "~1.0",
+        "phpdocumentor/reflection-docblock": "~3.0"
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/composer.json]: Why this one? Required by property info?

File: src/Symfony/Component/Serializer/Normalizer/ArrayDenormalizer.php (hunk 1 above)
Comment (dunglas) [inline src/Symfony/Component/Serializer/Normalizer/ArrayDenormalizer.php]: Brackets are used for the similar exception here https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php#L270

WSYT of standardizing messages with the rest of the code base in another PR?

File: src/Symfony/Component/Serializer/Tests/Normalizer/ObjectNormalizerTest.php (hunk 5 above)
Comment (dunglas) [inline src/Symfony/Component/Serializer/Tests/Normalizer/ObjectNormalizerTest.php]: oops.

File: src/Symfony/Component/Serializer/composer.json (hunk 6 above)
Comment (dunglas) [inline src/Symfony/Component/Serializer/composer.json]: Because I use the `ReflectionExtractor` in a test (and it's a soft dependency of PropertyInfo).

Theoretically, it should also work with PropertyInfo 2.8 (it's why I've not enforced `<3.1` in a `conflict` section). But 2.8 requires `phpdocumentor/reflection:^1.0.7` and 2.1+`phpdocumentor/reflection-docblock:^3.0`. And those 2 dependencies are in conflict...

So I just required PropertyInfo `^3.1` for the sake of simplicity.

File: src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php (hunk 2 above)
Comment (theofidry) [inline src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php]: if it's a standard in Symfony code base then http://symfony.com/doc/current/contributing/code/standards.html should be updated accordingly as well, the code sample L78 implies otherwise

**Comment by nicolas-grekas** [comment]
:+1:

File: src/Symfony/Component/Serializer/Tests/Normalizer/ObjectNormalizerTest.php [hunk 7]
```diff
@@ -740,6 +764,11 @@ class ObjectOuter
     private $inner;
     private $date;
 
+    /**
+     * @var ObjectInner[]
+     */
+    private $inners;
```
Comment (theofidry) [inline src/Symfony/Component/Serializer/Tests/Normalizer/ObjectNormalizerTest.php]: if we're missing this phpdoc, the serializer won't try to normalize `inner` before calling `setInners()` right?

File: src/Symfony/Component/Serializer/Tests/Normalizer/ObjectNormalizerTest.php [hunk 8]
```diff
@@ -759,6 +788,16 @@ public function getDate()
     {
         return $this->date;
     }
+
+    public function setInners(array $inners)
```
Comment (theofidry) [inline src/Symfony/Component/Serializer/Tests/Normalizer/ObjectNormalizerTest.php]: I would also support the case where we have:

```php
class ObjectOuter
{
  private $inners;

  /**
   * @parma ObjectInner[] $inners
   */
  public function setInners(array $inners)
  {
    $this->inners = $inners;
  }
}
```

File: src/Symfony/Component/Serializer/Tests/Normalizer/ObjectNormalizerTest.php (hunk 7 above)
Comment (dunglas) [inline src/Symfony/Component/Serializer/Tests/Normalizer/ObjectNormalizerTest.php]: If there is no type defined (through a PHPDoc or any other way supported by registered extractors), the old behavior (don't check anything) occurs.

File: src/Symfony/Component/Serializer/Tests/Normalizer/ObjectNormalizerTest.php (hunk 8 above)
Comment (dunglas) [inline src/Symfony/Component/Serializer/Tests/Normalizer/ObjectNormalizerTest.php]: This is out of scope of this PR. Support for such notations should be added in the PropertyInfo component.

Comment (theofidry) [inline src/Symfony/Component/Serializer/Tests/Normalizer/ObjectNormalizerTest.php]: fair enough

**Comment by theofidry** [comment]
👍

**Comment by fabpot** [comment]
Thank you @dunglas.
